Reusable CI workflow

weierophinney opened the pull request:
Defines a reusable CI workflow based on the standard one used in the majority of Laminas/Mezzio repositories. (The workflow is deliberately not in `.github/workflows/` to prevent it triggering in this repository.) Also documents how to consume it via the new `WORKFLOWS.md` file.

We will need to create the tag `ci-v1` after it has been tested and merged, to allow users to pin to a specific known-good version. Alternately, we can create a branch named `ci-v1` to track the specific workflow and allow making changes over time. If we do, however, the changes CANNOT require new inputs (though optional inputs would be allowed).

weierophinney committed:
feat: reusable CI workflow

Defines a reusable CI workflow based on the standard one used in the majority of Laminas/Mezzio repositories.
(The workflow is deliberately not in `.github/workflows/` to prevent it triggering in this repository.)
Also documents how to consume it via the new `WORKFLOWS.md` file.

We will need to create the tag `ci-v1` after it has been tested and merged, to allow users to pin to a specific known-good version.
Alternately, we can create a branch named `ci-v1` to track the specific workflow and allow making changes over time.
If we do, however, the changes CANNOT require new inputs (though optional inputs would be allowed).

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
fix: reusable workflows MUST be in .github/workflows directory

Noticed this when it didn't trigger for a test, and found information here:

- https://github.com/actions/runner/issues/1493

These WILL NOT trigger within this repo, as they only define `on: workflow_call`.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

internalsystemerror commented on WORKFLOWS.md:
```diff
@@ -0,0 +1,25 @@
+    branches:
+      - '[0-9]+.[0-9]+.x'
+      - 'refs/pull/*'
+      - 'renovate/*'
```
@Ocramius suggested just removing these entirely and having
```yml
on:
  pull_request:
  push:
    branches:
    tags:
```

weierophinney replied:
I can make that change. The reason we had the original set was to prevent runs on arbitrary branches unless they were used for pull requests, but since that never actually happens in practice, we could remove that stipulation.

weierophinney replied:
Updated!

Ocramius replied:
Suggestion: drop branch restrictions overall - no point in having them.

internalsystemerror commented on WORKFLOWS.md:
```diff
@@ -0,0 +1,25 @@
+
+jobs:
+  ci:
+    uses: laminas/.github/workflows/continuous-integration.yml@ci-v1
```
I'm just curious why the `ci-` prefix, do you expect to have different v1 tags for this or is it a best practice?

weierophinney replied:
Basically, so we can have multiple workflows in here (e.g., one for docs, one for release), and create tags for changes in each.

So, for example, if we made a change to the releases workflow that prompts a new major, that new major would not apply to the other two workflows.

The other option, of course, is to do these in separate branches; using a branch ref would allow them to change over time to pick up fixes or features.

I'm really not sure what would be the best solution, to be honest. Most of the time, I don't anticipate these changing very often, but when they do, we want to allow consumers to opt-in to the changes, instead of forcing users to opt-out by updating to specify an old ref.

internalsystemerror replied:
Ah I see now. In my mind using different files would have been all you need, but I didn't think that through properly.

weierophinney replied:
Updated!

internalsystemerror reviewed (commented):
Just a couple of comments to add, otherwise looks good. Would it be worth moving the `on` section into the shared workflow if that's going to be standard? I'm not too familiar with github CI.

weierophinney commented:
Example run: https://github.com/phly/phly-event-dispatcher/pull/9/checks

weierophinney commented:
> Would it be worth moving the `on` section into the shared workflow if that's going to be standard?

It'd be ideal, in fact... but reusable workflows apply at the jobs level, not at the workflow level, so we have to specify the `on` section in the workflow file in each repo. :upside_down_face:

weierophinney committed:
fix: remove specific branch references in example

Per @ise and @ocramius, we can likely safely build against any branch, not just release branches and PRs, since our policy is to send PRs from forks, not on the repo itself.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

internalsystemerror reviewed (approved):
LGTM, up to you how to handle the different workflows, I suspect tags may be less maintenance but that's just a guess.

weierophinney commented:
Closing; I'm going to push this to a new repo, so we can have a release branch and tag that are specific to the workflow, and capable of evolving separately from the CI actions it invokes.
